### ISSUE TITLE
Evaluation problem

### ISSUE THREAD

**Issue by Almo1982** [mcordts/cityscapesScripts#9]
Hi again Marius,

i was able to train my network as discussed in the other topic. I created result pictures and now i want to evaluate them.
I still have a problem with the evalPixelLevelSemanticLabeling.py

First it says that it cant find a prediction file...then i created a folder named "result" in my Cityscapes root directory. Then it says that it find multiple predictions for ground truth in '../../gtFine/val/lindau/lindau_000001_000019_gtFine_labelIds.png'! why? i have only two images in the result-folder...

Why is it always pointing to the gtFine/val...path no matter what i do??? For first experiment issues i only want to evaluate one prediction image. So i have one picture named 'lindau_000001_000019.png' and another picture named 'lindau_000001_000019_gtFine_labelIds.png' in the 'result' folder like described above.
I just want to compare this two pictures with each other and calculate the IoU etc....it took me already hours to step behind this, i was not able to do it...

What am i doing wrong here? 

Could you give my like a 'step by step' advice how to handle this? Like i said i would already be happy just to compare (evaluate) these two images before doing this with much more files.

Thank you

Stefan

**Comment by mcordts** [comment]
Hi Stefan,

the evaluation tool searches for ALL images in the validation set a matched prediction. In doing so, it iterates through all images in gtFine/val/* and tries to find a matched prediction. If it does not find one, it complains by printing the file name of the ground truth image. If it finds multiple it complains again. In your case the ground truth image is named "../../gtFine/val/lindau/lindau_000001_000019_gtFine_labelIds.png" and is printed since things went wrong. It found two matched predictions, i.e. "'lindau_000001_000019_gtFine_labelIds.png" and "lindau_000001_000019.png" in the result folder.

If you want to change this behaviour, you will need to modify the code or delete images that you don't want for evaluation from the ground truth folder "gtFine/val". Note that the IoU metric does not make much sense on a single image, since you need to have all classes at least once in your ground truth set.

Best,
Marius

**Comment by Almo1982** [comment]
Hi,

ok, got it so far, no more problems with reading the images. 
Now i face the next problem that says "Predicted image has multiple channels." Actually my prediction image has 3 channels but dont the instanceId and labelId- images have also 3 channels?!
My prediction image is colored, as i understand the script correctly the prediction image has to be grayscale with pixel values i defined for the labelTrainId-images (in my case 0-18 and 255), correct? --> so i have to transform these prediction image to this. 
The prediction image has to have just 1 channel then or 3? 

thank you!

Stefan

**Comment by Almo1982** [comment]
hi again,

i treid around a little bit...when i set if ( len(predictionNp.shape) != 3 ) in the script it proceeds (because i have three channels), why is it set to 2 by default? What is a 2 channel image???

i got my pixels in the prediction images now labeled between 0 and 33 like described in the labels.py. when i run the script now i get the error "ERROR: Unknown label with id (11, 11, 11)". Why is this? all labels between 0 and 33 should be known, or not? Or has it still to do with the 3 channels my images have?

thanks for your feedback!

Regards
Stefan

**Comment by Almo1982** [comment]
ok, proceeded a littel bit again, now i understand the part of "len(predictionNp.shape) != 2" in the Evaluation script. I will convert my prediction Image from shape (512, 256, 3) to grayscale what should result in (512, 256) that equals a length of the tuple of 2 then!

I hope this eliminates my other error "ERROR: Unknown label with id (11, 11, 11)", but it should! I guess code is looking for (11), not for (11, 11, 11), this might be (hopefully the last :-) ) Problem!

Regards
Stefan

**Comment by Almo1982** [comment]
ok, i finally got it to work now! 

i used cv2.COLOR_RGB2GRAY for my prediction images so that they are 1 channel now. What it still dont understand: the evalPixelLevelSemanticLabeling.py Script was still complaining that e.g. class (3,3,3) is not known. Then I also converted the ground truth *labelIds-images to 1 channel and it works! But the ground truth labelIds- images where directly downloaded from your homepage, so why are these pictures not 1 channel pictures from the beginning on? I just want to understand that...

And could you also answer my question what exactly the output of the evaluation script means:

-------------- ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------- 
              |  u   |  e   |  r   |  o   |  s   |  d   |  g   |  r   |  s   |  p   |  r   |  b   |  w   |  f   |  g   |  b   |  t   |  p   |  p   |  t   |  t   |  v   |  t   |  s   |  p   |  r   |  c   |  t   |  b   |  c   |  t   |  t   |  m   |  b   | Prior |
-------------- ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------ ------- 
    unlabeled | 0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.40   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.60   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00  0.0000 
  ego vehicle | 0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.96   0.01   0.00   0.00   0.03   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.01   0.00   0.00   0.00   0.00   0.00   0.00   0.00  0.0385 
   out of roi | 0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.32   0.02   0.00   0.00   0.45   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.14   0.00   0.07   0.00   0.00   0.01   0.00   0.00   0.00   0.00   0.00   0.00   0.00  0.0175 
       static | 0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.02   0.07   0.00   0.00   0.69   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.00   0.11   0.01   0.06   0.00   0.00   0.04   0.00   0.00   0.00   0.00   0.00   0.00   0.00  0.0147 
....(i didnt show all lines here) 
--> that are all 35 classes as far as i understand, but what does "Prior" and all the letters above mean in this case?

Furtheron is the outcoming "nIou" of the script the same like "iIoU" described in your documentation?

Thank you in advance!

Stefan

**Comment by Almo1982** [comment]
Hi Marius,

could you answer my last question? This is the last Thing not really clear for me. If this is answered then i am fine i think!

Thank you very much!

Stefan

**Comment by mcordts** [comment]
Hi Stefan,

the labelID-pngs that you can download from our website are all gray-value. Probably you applied some conversion there.

The script outputs a confusion matrix first. The letters on top are the first letters of the label names in the left column. The prior is the proportion of pixels with the respective ground truth label. The nIoU is the iIoU of the paper. Apparently, I used an n for normalized there instead of an i for instance. 

Best,
Marius

**Comment by Almo1982** [comment]
Hi Marius,

true, was my fault doing some conversion, that changed it from e.g. 11 to (11, 11, 11)...sorry about that but i am very new to the Topic of deep learning, Linux and Python. At the beginning there was more "try and error" then thinking a little bit sharper ;-) it's clear for me now and everything works now as expected.

Ok, understand with the confusion Matrix, the letters on top and nIoU, thanks!

Everything clear for me now, i already got some great results from a Conv Network trained with your dataset! 

Thank you again for your work on this dataset and your great Support in this Forum!!!

I will Close this issue!

Regards
Stefan
